pkg/view/router.go

```diff
@@ -1,74 +1,386 @@
+			log.Err(err).Msg("failed to get session")
+		}
+		flashes := session.Flashes()
+		err = session.Save(r, w)
+		if err != nil {
+			log.Err(err).Msg("failed to save session")
+		}
+
+		var fls []Flash
+		for _, flash := range flashes {
+			f, ok := flash.(Flash)
+			if !ok {
+				continue
+			}
+			fls = append(fls, f)
+		}
+
+		user, ok := getUser(r)
+		m := Model{
+			CSRFToken: csrf.TemplateField(r),
+			Flashes:   fls,
+		}
 		if !ok {
-			return Model{}
+			return m
 		}
-		return Model{
-			Email: user.Email,
+		m.Email = user.Email
+		return m
+	}
+
+	newFlash := func(w http.ResponseWriter, r *http.Request, ty FlashType, m string) {
+		// TODO breadchris how should these errors be handled?
+		session, err := sessionStore.Get(r, gorillaSessionName)
+		if err != nil {
+			log.Err(err).Msg("failed to get session")
+			return
+		}
+		session.AddFlash(Flash{
+			Type:    ty,
+			Message: m,
+		})
+		err = session.Save(r, w)
+		if err != nil {
+			log.Err(err).Msg("failed to save session")
 		}
 	}
 
 	r.Get("/", func(w http.ResponseWriter, r *http.Request) {
-		err := gv.Render(w, http.StatusOK, "pages/index", loadModel(r))
+		err := gv.Render(w, http.StatusOK, "pages/index", loadModel(r, w))
 		if err != nil {
 			http.Error(w, err.Error(), http.StatusInternalServerError)
 		}
 	})
 
 	r.Get("/connections", func(w http.ResponseWriter, r *http.Request) {
-		err := gv.Render(w, http.StatusOK, "pages/connections/index", loadModel(r))
+		user, ok := getUser(r)
+		if !ok {
+			http.Error(w, "User not found", http.StatusInternalServerError)
+			return
+		}
+		destinations, err := storageServices.Database.GetDestinations(r.Context(), user.ID)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		m := loadModel(r, w)
+		m.Connections = Connections{
+			Destinations: destinations,
+		}
+		err = gv.Render(w, http.StatusOK, "pages/connections/index", m)
 		if err != nil {
 			http.Error(w, err.Error(), http.StatusInternalServerError)
 		}
 	})
 
 	r.Get("/connections/new", func(w http.ResponseWriter, r *http.Request) {
-		err := gv.Render(w, http.StatusOK, "pages/connections/new", loadModel(r))
+		err := gv.Render(w, http.StatusOK, "pages/connections/upsert", loadModel(r, w))
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+	})
+
+	r.Post("/connections/upsert", func(w http.ResponseWriter, r *http.Request) {
+		user, ok := getUser(r)
+		if !ok {
+			http.Error(w, "User not found", http.StatusInternalServerError)
+			return
+		}
+
+		err := r.ParseForm()
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		id := r.Form.Get("id")
+		if id != "" {
+			http.Error(w, "Editing connections not yet supported", http.StatusBadRequest)
+			return
+		}
+
+		settings := map[string]any{}
+		name := r.Form.Get("name")
+
+		t := r.Form.Get("type")
+		switch t {
+		case "duckdb":
+			// XXX breadchris support file destination?
```

Comment:
Ignore the file destination for now. We never want a file destination for the hosted version, but self-hosted users might want this. Not sure how to differentiate these paths.